chore: roadmap 2025 (#1544)
defenseunicorns/pepr: docs/090_roadmap/README.md

    - Load test Pepr/KFC to identify bottlenecks and areas of improvement.
    - Ensure that Pepr/KFC can handle a large number of resources and events over a sustained period of time (nightly).

_2025 Roadmap_  
## Phase 1: Code Quality - Experimentation 

- **Q1**:

>   - **Eliminate Circular Dependencies, Complexity, and strongly type return statements**:

Author: add business value to bullets

>   - **Eliminate Circular Dependencies, Complexity, and strongly type return statements**:
>     - Turn on eslint settings to keep codebase clean, maintainable and standardized.

Author: ```suggestion
  - **Turn on eslint enforcement and configure settings and see no warnings**:
    - Eliminate circular dependencies, complexity, return statements, etc.
```

  - **Metric and Performance Baselining**:
    - Establish a baseline for performance and resource utilization metrics. Use this data to make informed decisions about the direction of the project in terms of Deno2
  - **Experimentation with Deno2**:
    - Experiment with Deno2 through Dash Days and see if it can be used in the project. Look into the performance improvements and new features that Deno2 brings to the table. Attempt to back the PeprStore through a Valkey Cache.
  - **OTEL Preparation**:
    - Come up with a plan to implement Open Telemetry. Specifically distributed tracing, metrics, logs and events. Use this data to make debugging easier from a UDS Core prespective. There will be documentation work on how to use an OTEL collector with a Pepr Module.

> 

Author: nightly release cadence

Author: Semantic release?

## Phase 2: Durable Storage for Metrics and Performance Tests / Transactional Pepr Store

- **Q2**:
  - **Professional Dashboard displaying metrics and performance tests originating from CI**:

>     - Begin engaging with potential contributors and users through social media, Kubernetes/Cloud Native Computing Foundation (CNCF) meetups, and other channels. Monitor and participate in our Slack channel.

Author: ```suggestion
```
delete me

something like GitHub Pages - d3 charting

>   - **Transactional Pepr Store Planning and Implementation**:

Author: Determine if it makes sense. Implement in Q3

   - Sus out details involved with having a transactional Pepr Store. What are the implications of this? What are the benefits? What are the drawbacks? What are the use-cases? What are the technologies that can be used to implement this?
  - **Deno2**:

>     - Depending on Q1, make a decision of whether or not to go with Deno2. If so, begin the migration process.

Author: Deno discovery to q2 -> Implementation to q3 if it happens

    - Ensure that Pepr/KFC can handle a large number of resources and events over a sustained period of time (nightly).

_2025 Roadmap_  
## Phase 1: Code Quality - Experimentation 

- **Q1**:
  - **Turn on eslint enforcement and configure settings and see no warnings**:
    - Eliminate circular dependencies, complexity, return statements, etc.
  - **Metric and Performance Baselining**:
    - Establish a baseline for performance and resource utilization metrics. Use this data to make informed decisions about the direction of the project in terms of Deno2
  - **OTEL Preparation**:
    - Come up with a plan to implement Open Telemetry. Specifically distributed tracing, metrics, logs and events. Use this data to make debugging easier from a UDS Core prespective. There will be documentation work on how to use an OTEL collector with a Pepr Module.
  - **Nightly Release**:
    - Establish a nightly release process. This will help us to catch bugs early and ensure that the project is always in a releasable state.

## Phase 2: Durable Storage for Metrics and Performance Tests / Transactional Pepr Store

- **Q2**:
  - **Semantic Versioning**:

>     - Establish a clear versioning scheme and adhere to it. This will help users to understand the impact of changes and make it easier to upgrade.

Reviewer: ```suggestion

```

We already use semver and it's a pre-req for Nightly releases that increment the patch version for us. I think we can delete this bullet point.

  - **Turn on eslint enforcement and configure settings and see no warnings**:
    - Eliminate circular dependencies, complexity, return statements, etc.
  - **Metric and Performance Baselining**:
    - Establish a baseline for performance and resource utilization metrics. Use this data to make informed decisions about the direction of the project in terms of Deno2
  - **OTEL Preparation**:
    - Come up with a plan to implement Open Telemetry. Specifically distributed tracing, metrics, logs and events. Use this data to make debugging easier from a UDS Core prespective. There will be documentation work on how to use an OTEL collector with a Pepr Module.
  - **Nightly Release**:
    - Establish a nightly release process. This will help us to catch bugs early and ensure that the project is always in a releasable state.

## Phase 2: Durable Storage for Metrics and Performance Tests / Transactional Pepr Store

- **Q2**:
  - **Semantic Versioning**:
    - Establish a clear versioning scheme and adhere to it. This will help users to understand the impact of changes and make it easier to upgrade.
  - **Professional Dashboard displaying metrics and performance tests originating from CI**:

>     - Begin engaging with potential contributors and users through social media, Kubernetes/Cloud Native Computing Foundation (CNCF) meetups, and other channels. Monitor and participate in our Slack channel.

Reviewer 2: ```suggestion
```
Copy pasta!

  - **Determine if a Transactional PeprStore makes sense**:

>    - Sus out details involved with having a transactional Pepr Store. What are the implications of this? What are the benefits? What are the drawbacks? What are the use-cases? What are the technologies that can be used to implement this?

Reviewer 2: ```suggestion
    - Sus out details involved with having a transactional Pepr Store. What are the implications of this? What are the benefits? What are the drawbacks? What are the use-cases? What are the technologies that can be used to implement this?
```

  - **Experimentation with Deno2**:
    - Experiment with Deno2 through Dash Days and see if it can be used in the project. Look into the performance improvements and new features that Deno2 brings to the table. 

## Phase 3: TBD

- **Q3**:
  - **Deno2 Implementation**:

>     - Move forward with migrating the project to Deno2. This phase will focus on adapting the codebase, conducting extensive testing, and creating comprehensive documentation to ensure a seamless transition.

Reviewer 2: ```suggestion
    - If determined to be advisable, move forward with migrating the project to Deno2 (starting with the kubernetes-fluent-client..?). This phase will focus on adapting the codebase, conducting extensive testing, and creating comprehensive documentation to ensure a seamless transition.
```
